code/modules/antagonists/space_dragon/carp_rift.dm

```diff
@@ -0,0 +1,239 @@
+/// The carp rift is currently charging.
+#define CHARGE_ONGOING 0
+/// The carp rift is currently charging and has output a final warning.
+#define CHARGE_FINALWARNING 1
+/// The carp rift is now fully charged.
+#define CHARGE_COMPLETED 2
+
+/datum/action/innate/summon_rift
+	name = "Summon Rift"
+	desc = "Summon a rift to bring forth a horde of space carp."
+	background_icon_state = "bg_default"
+	icon_icon = 'icons/mob/actions/actions_space_dragon.dmi'
+	button_icon_state = "carp_rift"
+
+/datum/action/innate/summon_rift/Activate()
+	var/datum/antagonist/space_dragon/dragon = owner.mind?.has_antag_datum(/datum/antagonist/space_dragon)
+	if(!dragon)
+		return
+	var/area/rift_location = get_area(owner)
+	if(!(rift_location.area_flags & VALID_TERRITORY))
+		to_chat(owner, span_warning("You can't summon a rift here! Try summoning somewhere secure within the station!"))
+		return
+	for(var/obj/structure/carp_rift/rift in dragon.rift_list)
```

Comment:
```suggestion
	for(var/obj/structure/carp_rift/rift as anything in dragon.rift_list)
```
We're only expecting carp rifts to be in this list, so we don't have to bother checking.